### PR TITLE
[FLINK-31414][checkpointing] exceptions in the alignment timer are ig…

### PR DESCRIPTION
**Pull request by zoltar9264**
…nored

## What is the purpose of the change

fix problem described in [FLINK-31414](https://issues.apache.org/jira/browse/FLINK-31414).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes, checkpointing)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by pnowojski** [comment]
It would be great to have a simple unit test of `org.apache.flink.streaming.runtime.io.checkpointing.BarrierAlignmentUtil#createRegisterTimerCallback` that shows that an exception is not hidden.

**Comment by zoltar9264** [comment]
Thanks @pnowojski , I will add unit later.

**Comment by zoltar9264** [comment]
Hi @pnowojski , I added a unit test for `BarrierAlignmentUtil#createRegisterTimerCallback`, please take a look.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/checkpointing/BarrierAlignmentUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+        BarrierAlignmentUtil.DelayableTimer delayableTimer =
+                BarrierAlignmentUtil.createRegisterTimerCallback(mailboxExecutor, timerService);
+
+        Duration delay = Duration.ofMinutes(10);
+
+        delayableTimer.registerTask(
+                () -> {
+                    // simulate IOException in checkpoint sync phase
+                    throw new IOException();
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/checkpointing/BarrierAlignmentUtilTest.java:58]: nit: I would throw a custom exception, like for example `org.apache.flink.runtime.operators.testutils.ExpectedTestException`. It's less likely to collide with some actual failure reported via `IOException`.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/checkpointing/BarrierAlignmentUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                },
+                delay);
+
+        timerService.advance(delay.toMillis());
+
+        assertThrows(
+                "BarrierAlignmentUtil.DelayableTimer should not hidden exception",
+                IOException.class,
+                () -> mailboxProcessor.runMailboxStep());
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/checkpointing/BarrierAlignmentUtilTest.java:67]: nit: If you use `org.apache.flink.util.ExceptionUtils#assertThrowable(...)` or `ExceptionUtils#findThrowable` the test would be more resilient (for example if the original exception will be wrapped on some layer.

**Review by pnowojski — APPROVED** [review]
Thanks! LGTM % two optional nits

Could you also squash the commits together?

**Comment by zoltar9264** [comment]
Thanks for your @pnowojski , I change the test as you commented and squash the commits.

**Review by pnowojski — APPROVED** [review]

**Pull request merged**